module/zfs/dsl_dataset.c

```diff
@@ -1955,18 +1961,25 @@ dsl_dataset_space(dsl_dataset_t *ds,
 		else
 			*availbytesp = 0;
 	}
+	rrw_enter(&ds->ds_bp_rwlock, RW_READER, FTAG);
 	*usedobjsp = BP_GET_FILL(&dsl_dataset_phys(ds)->ds_bp);
+	rrw_exit(&ds->ds_bp_rwlock, FTAG);
 	*availobjsp = DN_MAX_OBJECT - *usedobjsp;
 }
 
 boolean_t
 dsl_dataset_modified_since_snap(dsl_dataset_t *ds, dsl_dataset_t *snap)
 {
-	ASSERT(dsl_pool_config_held(ds->ds_dir->dd_pool));
+	dsl_pool_t *dp = ds->ds_dir->dd_pool;
+	uint64_t birth;
+
+	ASSERT(dsl_pool_config_held(dp));
```

Comment:
This will result in an unused variable warning on Linux for non-debug builds where the above ASSERT is optimized out.  This is the reason this code diverged in the first place.  We can either preserve the existing fix for this:

```
ASSERT(dsl_pool_config_held(ds->ds_dir->dd_pool));
```

Or alternately the `dsl_pool_t *dp` declaration can be wrapped with the ASSERTV macro so the variable is only defined for debug builds.

```
        ASSERTV(dsl_pool_t *dp = ds->ds_dir->dd_pool);
```